core/remote_state_verifier.go

```diff
@@ -0,0 +1,399 @@
+	chainHeadCh  chan ChainHeadEvent
+	chainHeadSub event.Subscription
+
+	// Channels
+	verifyCh  chan common.Hash
+	messageCh chan verifyMessage
+}
+
+func NewVerifyManager(blockchain *BlockChain, peers verifyPeers, allowUntrusted bool) *remoteVerifyManager {
+	verifiedCache, _ := lru.New(verifiedCacheSize)
+	vm := &remoteVerifyManager{
+		bc:            blockchain,
+		tasks:         make(map[common.Hash]*verifyTask),
+		peers:         peers,
+		verifiedCache: verifiedCache,
+		allowInsecure: allowUntrusted,
+
+		chainHeadCh: make(chan ChainHeadEvent, chainHeadChanSize),
+		verifyCh:    make(chan common.Hash, maxForkHeight),
+		messageCh:   make(chan verifyMessage),
+	}
+	vm.chainHeadSub = blockchain.SubscribeChainHeadEvent(vm.chainHeadCh)
+	return vm
+}
+
+func (vm *remoteVerifyManager) mainLoop() {
+	defer vm.chainHeadSub.Unsubscribe()
+
+	// load unverified blocks in a normalized chain and start a batch of verify task
+	header := vm.bc.CurrentHeader()
+	// Start verify task from H to H-11 if need.
+	vm.NewBlockVerifyTask(header)
+	pruneTicker := time.NewTicker(pruneInterval)
+	defer pruneTicker.Stop()
+	for {
+		select {
+		case h := <-vm.chainHeadCh:
+			vm.NewBlockVerifyTask(h.Block.Header())
+		case hash := <-vm.verifyCh:
+			vm.cacheBlockVerified(hash)
+			if task, ok := vm.tasks[hash]; ok {
+				delete(vm.tasks, hash)
+				remoteVerifyTaskCounter.Dec(1)
+				succeedRemoteVerifyTaskMeter.Mark(1)
+				succeedTaskExecutionTimer.Update(time.Since(task.startAt))
+				close(task.terminalCh)
+			}
+		case <-pruneTicker.C:
+			for hash, task := range vm.tasks {
+				if vm.bc.CurrentHeader().Number.Cmp(task.blockHeader.Number) == 1 &&
+					vm.bc.CurrentHeader().Number.Uint64()-task.blockHeader.Number.Uint64() > pruneHeightDiff {
+					delete(vm.tasks, hash)
+					remoteVerifyTaskCounter.Dec(1)
+					failedRemoteVerifyTaskMeter.Mark(1)
+					close(task.terminalCh)
+				}
+			}
+		case message := <-vm.messageCh:
+			if vt, ok := vm.tasks[message.verifyResult.BlockHash]; ok {
+				vt.messageCh <- message
+			}
+
+		// System stopped
+		case <-vm.bc.quit:
+			for _, task := range vm.tasks {
+				close(task.terminalCh)
+			}
+			return
+		case <-vm.chainHeadSub.Err():
+			return
+		}
+	}
+}
+
+func (vm *remoteVerifyManager) NewBlockVerifyTask(header *types.Header) {
+	for i := 0; header != nil && i <= maxForkHeight; i++ {
+		func(hash common.Hash) {
+			// if verified cache record that this block has been verified, skip.
+			if _, ok := vm.verifiedCache.Get(hash); ok {
+				return
+			}
+			// if there already has a verify task for this block, skip.
+			if _, ok := vm.tasks[hash]; ok {
+				return
+			}
+			diffLayer := vm.bc.GetTrustedDiffLayer(hash)
+			// if this block has no diff, there is no need to verify it.
+			var err error
+			if diffLayer == nil {
+				if diffLayer, err = vm.bc.GenerateDiffLayer(hash); err != nil {
+					log.Error("failed to get diff layer", "block", hash, "number", header.Number, "error", err)
+					return
+				} else {
```

Comment:
Why return here?

Reply:
I guess it should be `else if diffLayer == nil`

Reply:
fixed